kaldialign/__init__.py

```diff
@@ -1,38 +1,55 @@
 import math
-from typing import List, Tuple
 import random
+from typing import Dict, Iterable, List, Optional, Sequence, Tuple, TypeVar, Union
+
 import _kaldialign
 
+Symbol = TypeVar("Symbol")
+
 
-def edit_distance(a, b, sclite_mode=False):
+def edit_distance(
+    ref: Iterable[Symbol], hyp: Iterable[Symbol], sclite_mode: bool = False
+) -> Dict[str, Union[int, float]]:
     """
-    Compute the edit distance between sequences ``a`` and ``b``.
+    Compute the edit distance between sequences ``ref`` and ``hyp``.
     Both sequences can be strings or lists of strings or ints.
 
     Optional ``sclite_mode`` sets INS/DEL/SUB costs to 3/3/4 for
     compatibility with sclite tool.
 
-    Returns a dict with keys ``ins``, ``del``, ``sub``, ``total``,
-    which stand for the count of insertions, deletions, substitutions,
-    and the total number of errors.
+    Returns a dict with keys:
+    * ``ins`` -- the number of insertions (in ``hyp`` vs ``ref``)
+    * ``del`` -- the number of deletions (in ``hyp`` vs ``ref``)
+    * ``sub`` -- the number of substitutions
+    * ``total`` -- total number of errors
+    * ``ref_len`` -- the number of symbols in ``ref``
+    * ``err_rate`` -- the error rate  (total number of errors divided by ``ref_len``)
     """
-    int2sym = dict(enumerate(sorted(set(a) | set(b))))
+    int2sym = dict(enumerate(sorted(set(ref) | set(hyp))))
     sym2int = {v: k for k, v in int2sym.items()}
 
-    ai: List[int] = []
-    bi: List[int] = []
-    for sym in a:
-        ai.append(sym2int[sym])
+    refi: List[int] = []
+    hypi: List[int] = []
+    for sym in ref:
+        refi.append(sym2int[sym])
 
-    for sym in b:
-        bi.append(sym2int[sym])
+    for sym in hyp:
+        hypi.append(sym2int[sym])
 
-    return _kaldialign.edit_distance(ai, bi, sclite_mode)
+    ans = _kaldialign.edit_distance(refi, hypi, sclite_mode)
+    ans["ref_len"] = len(refi)
+    ans["err_rate"] = ans["total"] / len(refi)
```

Comment:
FYI: This line introduces the bug, that edit_distance doesn't work for an empty reference, it raises a ZeroDivisionError.

Reply:
right, I’ll fix that. We should probably return one of inf, nan, or None: WDYT?

Reply:
Thanks.

Hmm. How about mirroring the behavior of numpy?
```python
>>> np.array([1]) / np.array([0])
array([inf])
```

Reply:
Fixed in #21, releasing as v0.9.1